src/common/xo/choose-sr-for-each-vdis-modal/index.js

```diff
@@ -32,100 +27,57 @@ Collapsible.propTypes = {
 }
 
 @propTypes({
-  vdis: propTypes.array.isRequired,
-  predicate: propTypes.func
+  mainSrPredicate: propTypes.func,
+  onChange: propTypes.func.isRequired,
+  srPredicate: propTypes.func,
+  value: propTypes.objectOf(
+    propTypes.shape({
+      mainSr: propTypes.object,
+      mapVdisSrs: propTypes.object
+    })
+  ).isRequired,
+  vdis: propTypes.object.isRequired
 })
 export default class ChooseSrForEachVdisModal extends Component {
-  state = {
-    mapVdisSrs: {}
-  }
-
-  componentWillReceiveProps (newProps) {
-    if (
-      this.props.predicate !== undefined &&
-      newProps.predicate !== this.props.predicate
-    ) {
-      this.state = {
-        mainSr: undefined,
-        mapVdisSrs: {}
-      }
-    }
-  }
-
-  _onChange = props => {
-    this.setState(props)
-    this.props.onChange(props)
-  }
-
-  _onChangeMainSr = newSr => {
-    const oldSr = this.state.mainSr
-
-    if (oldSr == null || newSr == null || oldSr.$pool !== newSr.$pool) {
-      this.setState({
-        mapVdisSrs: {}
-      })
-    } else if (!newSr.shared) {
-      const mapVdisSrs = {...this.state.mapVdisSrs}
-      forEach(mapVdisSrs, (sr, vdi) => {
-        if (sr != null && newSr !== sr && sr.$container !== newSr.$container && !sr.shared) {
-          delete mapVdisSrs[vdi]
-        }
-      })
-      this._onChange({mapVdisSrs})
-    }
-
-    this._onChange({
-      mainSr: newSr
+  _onChange = newValues => {
+    this.props.onChange({
+      ...this.props.value,
+      ...newValues
     })
   }
 
-  _getSrPredicate = createSelector(
-    () => this.state.mainSr,
-    () => this.state.mapVdisSrs,
-    (mainSr, mapVdisSrs) => sr =>
-      isSrWritable(sr) &&
-      mainSr.$pool === sr.$pool &&
-      areSrsCompatible(mainSr, sr) &&
-      every(mapVdisSrs, selectedSr => selectedSr == null || areSrsCompatible(selectedSr, sr))
-  )
-
   render () {
-    const { props, state } = this
-    const { vdis } = props
+    const { props } = this
     const {
-      mainSr,
-      mapVdisSrs
-    } = state
-
-    const srPredicate = props.predicate || this._getSrPredicate()
+      mainSrPredicate = isSrWritable,
+      srPredicate = mainSrPredicate,
+      value: { mainSr, mapVdisSrs }
+    } = props
 
     return <div>
       <SelectSr
-        onChange={mainSr => props.predicate !== undefined
-          ? this._onChange({mainSr})
-          : this._onChangeMainSr(mainSr)
-        }
-        predicate={props.predicate || isSrWritable}
+        onChange={mainSr => this._onChange({ mainSr })}
```

Comment:
So, i need to create two `_onChange`. `_onChangeMainSr` and `_onChangeMapVdisSr`.